### PR TITLE
Check RegisterMetricAndTrackRateLimiterUsage error when starting BootstrapSigner & TokenCleaner controllers

### PR DESCRIPTION
**Pull request by frodenas**
**What this PR does / why we need it**:
Prevent `BootstrapSigner` and `TokenCleaner` controllers to start if `metrics.RegisterMetricAndTrackRateLimiterUsage` returns an error.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: complements #53571 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by deads2k** [comment]
It doesn't seem like a failure to add metrics should result not running the controller. I could go as far as a log message if you like.

**Comment by frodenas** [comment]
@deads2k 

1. Before #53571 was merged, an error registering a metric panicked the controller at startup time (see example at #52801), so this PR doesn't change the behavior in the sense that the controller is not started if there's a problem.
2. A log message can go unnoticed. When there's a problem and you need to look at your prometheus metrics, then you realize that you don't have metrics. IMO, I don't believe this is acceptable.
3. Errors registering a metric at the [metrics.RegisterMetricAndTrackRateLimiterUsage](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/metrics/util.go#L40) call usually happen because a contributor (not an end-user): a) created/updated a controller and specified an invalid `owner name`(ie not compliant with [prometheus metric name rules](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels) and what happened at [#52801](https://github.com/kubernetes/kubernetes/issues/52801#issuecomment-331286326)) b) created/updated a controller and used an `owner name` already registered. Better to catch up those errors during development/test time that not users experiencing this problem.

Thoughts?

**Comment by deads2k** [comment]
> Errors registering a metric at the metrics.RegisterMetricAndTrackRateLimiterUsage call usually happen because a contributor (not an end-user): a) created/updated a controller and specified an invalid owner name(ie not compliant with prometheus metric name rules and what happened at #52801) b) created/updated a controller and used an owner name already registered. Better to catch up those errors during development/test time that not users experiencing this proble

Ok.  That makes sense.

Please combine with https://github.com/kubernetes/kubernetes/pull/53591

File: pkg/controller/bootstrap/bootstrapsigner_test.go [hunk 1]
```diff
@@ -60,13 +64,13 @@ func newConfigMap(tokenID, signature string) *v1.ConfigMap {
 }
 
 func TestNoConfigMap(t *testing.T) {
-	signer, cl := newBootstrapSigner()
+	signer, cl, _ := newBootstrapSigner()
```
Comment (deads2k) [inline pkg/controller/bootstrap/bootstrapsigner_test.go]: don't ignore the error.

Comment (frodenas) [reply pkg/controller/bootstrap/bootstrapsigner_test.go]: ok

File: cmd/kube-controller-manager/app/bootstrap.go [hunk 2]
```diff
@@ -16,20 +16,32 @@ limitations under the License.
...
 
 func startBootstrapSignerController(ctx ControllerContext) (bool, error) {
-	go bootstrap.NewBootstrapSigner(
+	bsc, err := bootstrap.NewBootstrapSigner(
 		ctx.ClientBuilder.ClientGoClientOrDie("bootstrap-signer"),
 		bootstrap.DefaultBootstrapSignerOptions(),
-	).Run(ctx.Stop)
+	)
+	if err != nil {
+		return false, fmt.Errorf("error creating bootstrap signer controller: %v", err)
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/bootstrap.go]: `return true, ....`

File: cmd/kube-controller-manager/app/bootstrap.go [hunk 3]
```diff
@@ -16,20 +16,32 @@ limitations under the License.
...
 
 func startTokenCleanerController(ctx ControllerContext) (bool, error) {
-	go bootstrap.NewTokenCleaner(
+	tcc, err := bootstrap.NewTokenCleaner(
 		ctx.ClientBuilder.ClientGoClientOrDie("token-cleaner"),
 		bootstrap.DefaultTokenCleanerOptions(),
-	).Run(ctx.Stop)
+	)
+	if err != nil {
+		return false, fmt.Errorf("error creating token cleaner controller: %v", err)
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/bootstrap.go]: `return true, ....`

File: cmd/kube-controller-manager/app/core.go [hunk 4]
```diff
@@ -256,7 +256,9 @@ func startResourceQuotaController(ctx ControllerContext) (bool, error) {
 		Registry:                  generic.NewRegistry(quotaConfiguration.Evaluators()),
 	}
 	if resourceQuotaControllerClient.CoreV1().RESTClient().GetRateLimiter() != nil {
-		metrics.RegisterMetricAndTrackRateLimiterUsage("resource_quota_controller", resourceQuotaControllerClient.CoreV1().RESTClient().GetRateLimiter())
+		if err := metrics.RegisterMetricAndTrackRateLimiterUsage("resource_quota_controller", resourceQuotaControllerClient.CoreV1().RESTClient().GetRateLimiter()); err != nil {
+			return false, err
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/core.go]: `return true, ....`

File: pkg/controller/bootstrap/bootstrapsigner_test.go [hunk 5]
```diff
@@ -60,13 +64,19 @@ func newConfigMap(tokenID, signature string) *v1.ConfigMap {
 }
 
 func TestNoConfigMap(t *testing.T) {
-	signer, cl := newBootstrapSigner()
+	signer, cl, err := newBootstrapSigner()
+	if err != nil {
+		t.Errorf("Error creating BootstrapSigner: %v", err)
```
Comment (deads2k) [inline pkg/controller/bootstrap/bootstrapsigner_test.go]: `t.Fatal`.  the rest of the test won't work.

**Comment by deads2k** [comment]
minor comments.  squash, then lgtm

**Comment by frodenas** [comment]
@deads2k ok, will update it. Question: #53593, #53594, #53595. #53643, #53646 and #53648 are related. Do you want them combined in this PR (similar to #53591)? If yes, I'll pick up and fix those commits with your comments in this PR.

File: cmd/kube-controller-manager/app/core.go (hunk 4 above)
Comment (frodenas) [reply cmd/kube-controller-manager/app/core.go]: Why `true` when some lines below is returning `false`? https://github.com/frodenas/kubernetes/blob/04d400c29ba2b5588ceb548473e60e0ad2ea7721/cmd/kube-controller-manager/app/core.go#L266

Also, this `bool` indicates if the controller has been [started](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L478), and it's failing before the `Run`, so it should return `false` right?

**Comment by deads2k** [comment]
> @deads2k ok, will update it. Question: #53593, #53594, #53595. #53643, #53646 and #53648 are related. Do you want them combined in this PR (similar to #53591)? If yes, I'll pick up and fix those commits with your comments in this PR.

Yeah, gather them all here.

File: cmd/kube-controller-manager/app/core.go (hunk 4 above)
Comment (deads2k) [reply cmd/kube-controller-manager/app/core.go]: > Also, this bool indicates if the controller has been started, and it's failing before the Run, so it should return false right?

The bool indicates that it was enabled.  Some controllers have secondary enablement criteria (like API availability) that they own inside their own start methods.

File: cmd/kube-controller-manager/app/core.go [hunk 6]
```diff
@@ -68,8 +68,7 @@ func startServiceController(ctx ControllerContext) (bool, error) {
 		ctx.Options.ClusterName,
 	)
 	if err != nil {
-		glog.Errorf("Failed to start service controller: %v", err)
-		return false, nil
+		return true, fmt.Errorf("error creating Service controller: %v", err)
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/core.go]: This one shouldn't fail.  I don't recall the details, but this was made a glog for some reason.  It lives like this as a legacy.

**Comment by deads2k** [comment]
/lgtm

but I'm guessing something is double registering....

File: cmd/kube-controller-manager/app/core.go (hunk 6 above)
Comment (frodenas) [reply cmd/kube-controller-manager/app/core.go]: let me revert back this change and I will add a comment

**Comment by frodenas** [comment]
@deads2k PTAL, had to rebase and updated a new [test](https://github.com/kubernetes/kubernetes/tree/master/test/integration/daemonset) added upstream. Thanks.

**Comment by deads2k** [comment]
hanging out because of pkg/util/metrics.  That lives in service this.  tagging.

**Pull request merged**